src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -149,6 +145,8 @@ public sealed class Builder
 
             private Exception? _exception = null;
 
+            private bool _isCompacted = true;
```

Comment:
>_isCompacted [](http://example.com/codeflow?start=25&length=12)

Do we need to track this value as items are added to the `Builder` or could it be calculated directly from `_states` in `ToImmutableAndFree()`?

Reply:
We can calculate it, bnut we'd have to iterate the _states to do so. It seemed more efficient to track it as we go, especiallty when _states has a large number of entries.

---
In reply to: [626253527](https://github.com/dotnet/roslyn/pull/53147#discussion_r626253527) [](ancestors = 626253527)

Reply:
Aren't we setting `_isCompacted` once (incrementally) for each entry currently? If instead, we calculate `_isCompacted` at the point where we have the complete set of `_states`, it would reduce the chance that we missed an entry, and would allow short-circuit with `_isCompacted = _states.Any(...);`.